fix(template-compiler): omit api_key for key outside iteration

nolanlawson opened the pull request:
## Details

Partially addresses #2697.

The idea is that if a template is like this:

```html
<template>
  <x-foo key={myKey}></x-foo>
</template>
```

Then we should _not_ render it with `api_key`, but instead ignore the `key`.

**How we currently render it:**

```js
function tmpl($api, $cmp) {
  const { k: api_key, c: api_custom_element } = $api;
  return [
    api_custom_element("x-foo", _xFoo, { key: api_key(0, $cmp.myKey) }),
  ];
}
```

**This PR:**

```js
function tmpl($api) {
  const { c: api_custom_element } = $api;
  return [api_custom_element("x-foo", _xFoo, { key: 0 })];
}
```

This PR also emits a warning if we find a `key` outside of an iteration.

## Does this pull request introduce a breaking change?

<!--
    Any change that can cause downstream consumers to fail qualifies as a breaking change.
    
    Examples:
        - Removing the code for a deprecated API.
        - Adding a new restriction to the compiler which might result in a compilation failure for existing code.
        - Changing the return type of a function in a non-backward compatible fashion.

    Remove the incorrect item for the list. 
-->
* ✅ No, it does not introduce a breaking change.

<!-- If yes, please describe the impact and migration path for existing applications. -->

## Does this pull request introduce an observable change?

<!--
    Observable changes are internal changes that can be observed by downstream consumers. 
    Such changes don't qualify as breaking changes because they don't impact any publicly defined 
    APIs.

    Examples:
        - Fixing a bug.
        - Changing the invocation timing of a callback, for a callback that has no invocation timing
          guarantee.

    Remove the incorrect item from the list. 
-->
* ✅ No, it does not introduce an observable change.

Almost everyone using a `key` outside of an iteration is probably doing it as a mistake, and even if they are, it should only impact the internals of our vnode-diffing engine.

<!-- If yes, please describe the anticipated observable changes. -->

## GUS work item
W-10720220

nolanlawson commented on packages/@lwc/engine-core/src/framework/rendering.ts:
```diff
@@ -701,15 +701,8 @@ function updateStaticChildren(c1: VNodes, c2: VNodes, parent: ParentNode) {
-                        //  iteration, leading to a situation where static children have different keys.
-                        unmount(n1, parent, true);
-                        mount(n2, parent, anchor);
-                    }
```
This code is now unnecessary! And we have a test to prove it.

Previous discussion: https://github.com/salesforce/lwc/pull/2677#discussion_r808534831

nolanlawson commented:
/nucleus test

pmdartus commented on packages/@lwc/template-compiler/src/parser/index.ts:
```diff
@@ -763,7 +764,19 @@ function applyKey(
+            );
+        }
+
+        element.directives.push(
+            ast.keyDirective(keyAttribute.value, keyAttribute.location, parentIsIteration)
+        );
```
I don't think it makes sense to keep the `key` directive as part of the AST if it doesn't carry any semantic.

nolanlawson replied:
You're right. Fixed.

pmdartus commented on packages/@lwc/template-compiler/src/parser/index.ts:
```diff
@@ -763,7 +764,19 @@ function applyKey(
+        const parentIsIteration = !!(forOfParent || forEachParent);
+
+        if (!parentIsIteration) {
+            ctx.warnAtLocation(
```
You can use `warnOnNode` instead of `warnAtLocation` this would avoid having to pass the `parse5ElmLocation`.

```suggestion
            ctx.warnOnNode(
```

nolanlawson replied:
Thanks for the tip! Fixed.

nolanlawson commented:
/nucleus test

nolanlawson commented:
/nucleus test

nolanlawson commented:
`lwc-platform` downstream test failing due to Komaci test. Will look into it.

nolanlawson commented:
So, it turns out there was a big use case we missed here – inline iterators and `for:each`:

```html
<x-foo iterator:x={items} key={x.value} index={x.index} last={x.last} first={x.first}></x-foo>
```

```html
<x-foo for:each={items} for:item="item" key={item.id}></x-foo>
```

In this case, the `key` can appear directly on the same element with `for:each` or `iterator:x`. We weren't handling this situation properly. I added a test and fixed this.

nolanlawson commented on packages/@lwc/template-compiler/src/parser/index.ts:
```diff
@@ -1142,14 +1141,14 @@ function isInIteration(ctx: ParserCtx): boolean {
-function isInIteratorElement(ctx: ParserCtx, parent: ParentNode): boolean {
-    return !!(getForOfParent(ctx, parent) || getForEachParent(ctx));
+function isInIteratorElement(ctx: ParserCtx): boolean {
+    return !!(getForOfParent(ctx) || getForEachParent(ctx));
```
@jmsjtu @pmdartus This part was really tricky – we had slightly different logic for `getForOfParent` and `getForEachParent`. It was causing `getForOfParent` not to work with inline iterators, because I was ignoring the `key` in that case.

I guess this didn't matter much before, because these tests were just looking for an error case (keys should not reference the index, or iteration missing a key entirely).

In my fix, we don't care if the `for:each` or `iterator:x` is on the same node or the parent node. I'd appreciate if you took a look to make sure I didn't miss something.

nolanlawson replied:
...Just confirmed that it's possible to use an `index` as a `key` if the `iterator` is inline. My PR fixes this.

I can open it as a separate PR if you prefer. This one is getting pretty complex.

nolanlawson replied:
Opened a separate PR: https://github.com/salesforce/lwc/pull/2703

nolanlawson commented on packages/integration-karma/test/rendering/iteration/index.spec.js:
```diff
@@ -0,0 +1,59 @@
+            });
+        });
+    });
+});
```
This test passes in `master`. On 236 (v2.5.13) and 234 (v2.2.12), it throws an error:

    Error: Invariant Violation: Compiler should produce html functions that always return an array.

The reason for the error is that the `tmpl` function is returning `api_custom_element` rather than `api_iterator`. So it looks like previously we didn't actually support inline iterators, only inline `for:each`.
